src/instructions.rs

```
@@ -0,0 +1,185 @@
+fn from_bits(value: u16, size: u16, offset: u16) -> u16 {
+    (value >> offset) & ((1 << size) - 1)
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum OperationCode {
+    Branch,
+    Add,
+    Load,
+    Store,
+    JumpRegister,
+    And,
+    LoadRegister,
+    StoreRegister,
+    /// Unused
+    Rti,
+    Not,
+    LoadIndirect,
+    StoreIndirect,
+    Jump,
+    /// Unused
+    Reserved,
+    LoadEffectiveAddress,
+    ExecuteTrap,
+}
+
+impl OperationCode {
+    fn from_u16(value: u16) -> Self {
+        match value {
+            0 => Self::Branch,
+            1 => Self::Add,
+            2 => Self::Load,
+            3 => Self::Store,
+            4 => Self::JumpRegister,
+            5 => Self::And,
+            6 => Self::LoadRegister,
+            7 => Self::StoreRegister,
+            8 => Self::Rti,
+            9 => Self::Not,
+            10 => Self::LoadIndirect,
+            11 => Self::StoreIndirect,
+            12 => Self::Jump,
+            14 => Self::LoadEffectiveAddress,
+            15 => Self::ExecuteTrap,
+            _ => Self::Reserved, // We create noops for any other operation code
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum ConditionFlag {
+    Positive = 1 << 0,
+    Zero = 1 << 1,
+    Negative = 1 << 2,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Register {
+    R0,
+    R1,
+    R2,
+    R3,
+    R4,
+    R5,
+    R6,
+    R7,
+    PC, /* program counter */
+    Cond,
+}
+/// Number of available registers
+pub const REGISTER_COUNTER: usize = 10;
+
+impl Register {
+    fn from_u16(value: u16) -> Self {
+        match value {
+            0 => Self::R0,
+            1 => Self::R1,
+            2 => Self::R2,
+            3 => Self::R3,
+            4 => Self::R4,
+            5 => Self::R5,
+            6 => Self::R6,
+            7 => Self::R7,
+            8 => Self::PC,
+            _ => Self::Cond,
+        }
+    }
```

I think that creating the `Self::Cond` with any number that is not in the range 0..8 is strange. Maybe you could make the function to return a Result so you can return an Err

---

For now is todo!(), later I will consider adding something that blows up on wrong numbers